app/src/main/java/pl/ppiwd/exerciseanalyst/activities/TrainingDataActivity.java

```diff
@@ -0,0 +1,254 @@
+        metaMotionServiceConnection = new MetaMotionServiceConnection();
+        bindService(serviceIntent, metaMotionServiceConnection, Context.BIND_AUTO_CREATE);
+    }
+
+    private void finishSession() {
+        unbindMetaMotionService();
+        stopService(deviceService);
+
+        startStopButton.setOnClickListener(view -> handleSession());
+        startStopButton.setText(R.string.start_training);
+        activitySpinner.setEnabled(true);
+        repetitionsPicker.setEnabled(true);
+    }
+
+    private void unbindMetaMotionService() {
+        if (metaMotionServiceConnection != null) {
+            Log.i("DataCollectionActivity()", "unbindMetaMotionService() unbinding");
+            unbindService(metaMotionServiceConnection);
+            metaMotionServiceConnection = null;
+        }
+    }
+
+    private void sendDataToServer() {
+        DialogInterface.OnClickListener dialogClickListener = (dialog, clickedButton) -> {
+            String trainingType = getIntent().getStringExtra(Constants.TRAINING_TYPE);
+            if (clickedButton == DialogInterface.BUTTON_POSITIVE) {
+                if (trainingType.equals(Constants.TRAINING_DATA_GATHERING)) {
+                    serverConnection.sendModelTrainingData(BuildConfig.GATHER_DATA_URL);
+                } else if (trainingType.equals(Constants.TRAINING_REGULAR)) {
+                    //todo
+                    // This is the method which sends data to url which should recognize activity.
+                    // Sent data shouldn't contain info about repetitions and activity.
+                    // Currently it does and it should be changed.
+                    // (Possibly server will simply not read additional fields and it might work fine)
+                    serverConnection.sendModelTrainingData(BuildConfig.RECOGNIZE_DATA_URL);
+                }
+            }
+        };
+        AlertDialog.Builder builder = new AlertDialog.Builder(this);
+        builder.setMessage("Training finished. Do you want to save gathered data?")
+                .setPositiveButton("Yes", dialogClickListener)
+                .setNegativeButton("No", dialogClickListener)
+                .show();
+    }
+
+    //Permissions handling below...
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        // NOTE: delegate the permission handling to generated method
+        TrainingDataActivityPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
+    }
+
+    @OnPermissionDenied(Manifest.permission.ACCESS_FINE_LOCATION)
+    public void onBackgroundLocationDenied() {
+        Toast.makeText(this, "Background location denied", Toast.LENGTH_LONG).show();
+    }
+
+    @OnNeverAskAgain(Manifest.permission.ACCESS_FINE_LOCATION)
+    public void onCameraNeverAskAgain() {
```

Comment:
Now I see this method name was wrong the entire time :) it could be changed if there is an opportunity to do so  
"onBackgroundLocationNeverAskAgain"